dask/dataframe/core.py

```diff
@@ -1173,24 +1173,50 @@ def reduction(x, chunk, aggregate, token=None):
 def concat(dfs):
     """ Concatenate dataframes along rows
 
-    Currently only supports unknown divisions
+    - If divisions are known, concatenate every divisions each by each.
+    - If divisions are unknown, concatenate to the last
+    - Concatenating DataFrames mixed with known and unknown divisions are not
+      supported.
     """
-    if any(df.known_divisions for df in dfs):
-        # For this to work we need to add a final division for "maximum element"
-        raise NotImplementedError("Concat can't currently handle dataframes"
-                " with known divisions")
-    name = 'concat-' + tokenize(*dfs)
-    dsk = dict()
-    i = 0
-    for df in dfs:
-        for key in df._keys():
-            dsk[(name, i)] = key
-            i += 1
+    if not isinstance(dfs, list):
+        dfs = [dfs]
+    if len(dfs) == 0:
+        raise ValueError('Input must be a list')
+
+    if all(df.known_divisions for df in dfs):
+        from .multi import _maybe_align_partitions
+        dfs = _maybe_align_partitions(dfs)
+        name = 'concat-{0}'.format(tokenize(*dfs))
+
+        dsk = dict()
+        for i in range(dfs[0].npartitions):
+            dsk[(name, i)] = (_concat, (list, [(df._name, i) for df in dfs]))
+
+        dasks = [df.dask for df in dfs]
 
-    divisions = [None] * (i + 1)
+        columns = toolz.concat([df.columns for df in dfs])
+        columns = sorted(unique(columns))
```

Comment:
Is this exactly how Pandas does it?  Historically I've actually done the pandas computation on empty dataframes to collect the appropriate metadata.

    columns = pd.concat([pd.DataFrame([], columns=df.columns) for df in dfs]).columns

This was both because I didn't know the policy taken by Pandas and so that it would track future changes.